10_nwis_pull/src/nwis_pull.R

```diff
@@ -0,0 +1,109 @@
+#### pull ####
+
+# prepare a plan for downloading (from NWIS) and posting (to GD) one data file
+# per state
+plan_nwis_pull <- function(partitions, service) {
+
+  folders <- list(
+    tmp='10_nwis_pull/tmp',
+    out='10_nwis_pull/out',
+    log='10_nwis_pull/log')
+  
+  # after all wanted data have been pulled, this function will be called but
+  # doesn't need to create anything much, so just return NULL
+  # isolate the partition info for just one task
+  partitions_name <- deparse(substitute(partitions))
+  partition <- scipiper::create_task_step(
+    step_name = 'partition',
+    target_name = function(task_name, step_name, ...) {
+      sprintf('%s_partition_%s', service, task_name)
+    },
+    command = function(task_name, ...) {
+      sprintf("filter_partitions(partitions = %s, I('%s'))", partitions_name, task_name)
+    }
+  )
+
+  # download from NWIS, save, and create .ind file promising the download is
+  # complete, with extension .tind ('temporary indicator') because we don't want
+  # scipiper to make a build/status file for it
+
+  download <- scipiper::create_task_step(
+    step_name = 'download',
+    target_name = function(task_name, step_name, ...) {
+      file.path(folders$tmp, sprintf('%s_%s.rds', service, task_name))
+    },
+    command = function(steps, ...) {
+      paste(
+        "get_nwis_data(",
+        "data_file=target_name,",
+        paste0("service = I('", service, "'),"),
+        sprintf("partition=%s,", steps$partition$target_name),
+       "nwis_pull_params = nwis_pull_parameters)",
+        sep="\n      ")
+    }
+  )
+
+  # put the steps together into a task plan
+  task_plan <- scipiper::create_task_plan(
+    task_names=sort(partitions$PullTask),
+    task_steps=list(partition, download),
+    final_steps=c('download'),
+    add_complete=FALSE,
+    ind_dir=folders$tmp)
+  
+  return(task_plan)
+
+}
+
+create_nwis_pull_makefile <- function(makefile, task_plan, final_targets) {
+
+
+  create_task_makefile(
+    makefile=makefile, task_plan=task_plan,
+    include = c('10_nwis_pull.yml'),
+    sources = c('10_nwis_pull/src/nwis_pull.R', '10_nwis_pull/src/nwis_combine_functions.R'),
+    packages=c('dplyr', 'dataRetrieval', 'feather', 'scipiper', 'yaml', 'stringr'),
+    file_extensions=c('ind','feather'), finalize_funs = 'combine_nwis_data', final_targets = final_targets)
+}
+
+# --- functions used in task table ---
+
+# read the partitions file and pull out just those sites that go with a single
+# PullTask 
+filter_partitions <- function(partitions, pull_task) {
+  
+  these_partitions <- dplyr::filter(partitions, PullTask==pull_task, count_nu > 0)
+  
+  return(these_partitions)  
+}
+
+# pull a batch of NWIS observations, save locally, return .tind file
+get_nwis_data <- function(data_file, partition, nwis_pull_params, service, verbose = TRUE) {
+  
+  nwis_pull_params$service <- service
+  nwis_pull_params$site <- partition$site_no
+  
+  dataRetrieval::setAccess(access = 'internal')
```

Comment:
will this fail if someone from outside USGS tries it? Or if I am not on the network? May need to put in a `tryCatch` and a note somewhere that USGSers on VPN will get different results.

Reply:
No, this will not fail. Is there some way to test if you're on VPN?

Reply:
Uhhhh not sure ... was just digging in the `dataRetrieval::setAccess`, but don't see much in there. Will ask in DS channel

Reply:
This is definitely something I would forget to verify myself, especially now that I am at home and not just automatically on the Network like I am at work.

Reply:
Should we keep it public data only?

Reply:
Yes, sounds like public data only is the way people are leaning